LOEUF.pm

```diff
@@ -0,0 +1,171 @@
+ https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7334197/
+
+
+ LOEUF scores can be downloaded from
+ https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7334197/bin/41586_2020_2308_MOESM4_ESM.zip
+
+ For GRCh37:
+
+ These files can be tabix-processed by:
+ Unzip 41586_2020_2308_MOESM4_ESM.zip
+ cd supplement
+ zcat supplementary_dataset_11_full_constraint_metrics.tsv.gz | (head -n 1 && tail -n +2  | sort -t$'\t' -k 76,76 -k 77,77n ) | bgzip -c > loeuf_dataset.tsv.gz
+ tabix -f -S 1 -s 76 -b 77 -e 77 loeuf_dataset.tsv.gz
+
+ The tabix utility must be installed in your path to use this plugin.
+
+=cut
+
+package LOEUF;
+
+use strict;
+use warnings;
+
+use Bio::EnsEMBL::Utils::Sequence qw(reverse_comp);
+
+use Bio::EnsEMBL::Variation::Utils::BaseVepTabixPlugin;
+
+use base qw(Bio::EnsEMBL::Variation::Utils::BaseVepTabixPlugin);
+
+use Scalar::Util qw(looks_like_number);
+
+
+sub new {
+  my $class = shift;
+
+  my $self = $class->SUPER::new(@_);
+
+  $self->expand_left(0);
+  $self->expand_right(0);
+
+  my $param_hash = $self->params_to_hash();
+
+  die("ERROR: LOEUF file not provided or not found!\n") unless defined($param_hash->{file}) && -e $param_hash->{file};
+  $self->add_file($param_hash->{file});
+
+  if(defined($param_hash->{match_by})) {
+    my $match_by = $param_hash->{match_by};
+    $self->{match_by} = $match_by;
+  }
+
+  else{
+    die("ERROR: Argument 'match_by' is undefined");
+  }
+
+  return $self;
+}
+
+sub feature_types {
+  return ['Transcript'];
+}
+
+sub get_header_info {
+  return { LOEUF => 'Loss-of-function observed/expected upper bound fraction'};
+}
+
+sub run {
+  my ($self, $tva) = @_;
+
+  return {} if grep {$_->SO_term eq 'downstream_gene_variant' || $_->SO_term eq 'upstream_gene_variant'} @{$tva->get_all_OverlapConsequences};
+  
+  my $vf = $tva->variation_feature;
+  my $transcript = $tva->transcript;
+  my $end = $vf->{end};
+  my $start = $vf->{start};
+  ($start, $end) = ($end, $start) if $start > $end;
+  
+  if ($self->{match_by} eq 'transcript'){
+    my ($res) = grep {
+    $_->{transcript_id}  eq $transcript->stable_id;
+    } @{$self->get_data($vf->{chr}, $start, $end)};
+    return $res ? $res->{result} : {};
+  }
+
+  elsif ($self->{match_by} eq 'gene'){
+    my @data = grep {
+    $_->{gene_id} eq $transcript->{_gene_stable_id}
+    } @{$self->get_data($vf->{chr}, $start, $end)};
+    ## Get max loeuf value for the gene
+    my $max_loeuf_score = 0;
+    foreach (@data){
+      if (looks_like_number($_->{result}->{LOEUF})){
+        if ($_->{result}->{LOEUF} > $max_loeuf_score){
+          $max_loeuf_score = $_->{result}->{LOEUF};
+        }
+      }
+    }
+    return $max_loeuf_score ? {LOEUF => $max_loeuf_score} : {};
+  }
+
+  else{
+    return {};
+  }
+
+}
+
+sub parse_data {
+  my ($self, $line) = @_;
+  my @values = split /\t/, $line;
+  my ($transcript_id, $oe_lof_upper, $gene_id, $chromosome, $start_position, $end_position ) = @values[1,30,64,75,76,77];
```

Comment:
Suggest checking the file headers are as expected in sub new

Reply:
Sure, will add this check!